File: 792-number-of-matching-subsequences/792-number-of-matching-subsequences.py
class Solution:
    def numMatchingSubseq(self, s: str, words: List[str]) -> int:
        res = 0
        heads = []
        for i in range(26):
            heads.append([])
        
        for w in words:
            heads[ord(w[0]) - ord('a')].append((w, 0))
        
        for c in s:
            cWords = heads[ord(c) - ord('a')]
            # print(cWords)
            nextWordsSameChar = []
            for cWord in cWords:
                if cWord[1] + 1 == len(cWord[0]):
                    res += 1
                else:
                    nextWord = (cWord[0], cWord[1] + 1)
                    # print(nextWord)
                    if nextWord[0][nextWord[1]] == c:
                        nextWordsSameChar.append(nextWord)
                    else:
                        heads[ord(nextWord[0][nextWord[1]]) - ord('a')].append(nextWord)
            heads[ord(c) - ord('a')] = nextWordsSameChar
        
        return res
        

File: 792-number-of-matching-subsequences/792-number-of-matching-subsequences.java
class Solution
{
    public int numMatchingSubseq(String s, String[] words)
    {
        int res = 0;
        List<ArrayDeque<Pair>> heads = new ArrayList<>();
        for (int i = 0; i < 26; i++)
            heads.add(new ArrayDeque<Pair>());
        
        for (String w : words)
            heads.get(w.charAt(0) - 'a').offer(new Pair(w, 0));
        
        for (char c : s.toCharArray())
        {
            // Get the list for this character
            ArrayDeque<Pair> cHeads = heads.get(c - 'a');
            
            // Need to grab this ahead of time, list
            // could have items added to it.
            int initialSize = cHeads.size();
            for (int i = 0; i < initialSize; i++)
            {
                Pair p = cHeads.poll();
                p.index += 1;
                if (p.index == p.word.length())
                    res++;
                else
                    heads.get(p.word.charAt(p.index) - 'a').offer(p);
            }
        }
        
        return res;
    }
}

class Pair
{
    String word;
    int index;
    
    public Pair(String w, int i)
    {
        this.word = w;
        this.index = i;
    }
}